examples/seismic/source.py

```diff
@@ -103,34 +105,39 @@ class PointSource(SparseTimeFunction):
         Represents the number of points in this source.
     """
 
-    def __new__(cls, **kwargs):
-        name = kwargs.pop('name')
-        grid = kwargs.pop('grid')
-        time_range = kwargs.pop('time_range')
-        time_order = kwargs.pop('time_order', 2)
-        p_dim = kwargs.pop('dimension', Dimension(name='p_%s' % name))
-
-        coordinates = kwargs.pop('coordinates', kwargs.pop('coordinates_data', None))
-        # Either `npoint` or `coordinates` must be provided
-        npoint = kwargs.pop('npoint', None)
-        if npoint is None:
-            if coordinates is None:
-                raise TypeError("Need either `npoint` or `coordinates`")
-            npoint = coordinates.shape[0]
-
-        # Create the underlying SparseTimeFunction object
-        obj = SparseTimeFunction.__new__(cls, name=name, grid=grid,
-                                         dimensions=(grid.time_dim, p_dim),
-                                         npoint=npoint, nt=time_range.num,
-                                         time_order=time_order,
-                                         coordinates=coordinates, **kwargs)
-
-        obj._time_range = time_range._rebuild()
-
-        # If provided, copy initial data into the allocated buffer
-        data = kwargs.get('data')
-        if data is not None:
-            obj.data[:] = data
+    def __new__(cls, *args, **kwargs):
+        options = kwargs.get('options', {})
+        if cls in _SymbolCache:
```

Comment:
probably no need for _SymbolCache. I **think** the (right) way to achieve the same thing is more like [this](https://github.com/opesci/devito/blob/master/devito/types/basic.py#L472)

Reply:
to be more precise, I think we should use __init__, not __new__. This __new__ overriding looks hacky to me

Reply:
No it is needed, this is not `__init__` this `_new_` so it doesn't check the one in `basic`, doesn't work with just init

Reply:
W can't use init, these one have specific inputs and properties that disapear at any manipulation (subs, .dt,..) without the proper new

Reply:
I'm not sure I understand, perhaps I'll try this to see what's actually going on. And we can close an eye for this small PR, not an issue. But this is definitely a type of class that an user could want to write in application code, and we can't expect him/her to be able to know and deal with low-level things such as SymbolCache

Reply:
That's a good point

Reply:
Just like to add that if we're using `_SymbolCache` outside the devito package, it shouldn't be private(marked with an underscore)

Reply:
indeed we shouldn't